css/css-fonts/support/js/variation-sequences.js

```diff
@@ -0,0 +1,125 @@
+var baseChars = {
+    "emoji": "\u{1fae8}",
+    "cjk": "\u{8279}",
+    "math": "\u{2205}"
+};
+
+var variationSelectors = {
+    "emoji": ["\u{fe0e}", "\u{fe0f}"],
+    "cjk": ["", "\u{FE00}", "\u{FE01}", "\u{e0100}", "\u{e0101}",
+        "\u{e0102}"
+    ],
+    "math": ["", "\u{FE00}"]
+};
+
+var families = {
+    "emoji": ["ColorEmojiFont", "MonoEmojiFont",
+        "EmojiFontWithBaseCharOnly",
+        "sans-serif"
+    ],
+    "cjk": ["CJKFontWithVS", "CJKFontWithBaseCharOnly",
+        "sans-serif"
+    ],
+    "math": ["MathFontWithVS", "MathFontWithBaseCharOnly",
+        "sans-serif"
+    ]
+};
+
+var variationSequenceFamilies = new Map([
+    ["\u{1fae8}\u{fe0e}", "MonoEmojiFont"],
+    ["\u{1fae8}\u{fe0f}", "ColorEmojiFont"],
+    ["\u{8279}\u{fe00}", "CJKFontWithVS"],
+    ["\u{8279}\u{fe01}", "CJKFontWithVS"],
+    ["\u{8279}\u{e0100}", "CJKFontWithVS"],
+    ["\u{8279}\u{e0101}", "CJKFontWithVS"],
+    ["\u{8279}\u{e0102}", "CJKFontWithVS"],
+    ["\u{2205}\u{FE00}", "MathFontWithVS"]
+]);
+
+var baseCharFamilies = new Map([
+    ["\u{1fae8}", new Set(["MonoEmojiFont", "ColorEmojiFont",
+        "EmojiFontWithBaseCharOnly"
+    ])],
+    ["\u{8279}", new Set(["CJKFontWithVS",
+        "CJKFontWithBaseCharOnly"
+    ])],
+    ["\u{2205}", new Set(["MathFontWithVS",
+        "MathFontWithBaseCharOnly"
+    ])]
+]);
+
+const range = function*(l) {
+    for (let i = 0; i < l; i += 1) yield i;
+}
+const isEmpty = arr =>
+    arr.length === 0;
+
+const permutations =
+    function*(a) {
+  const r = arguments[1] || [];
+  if (isEmpty(a))
+    yield r;
+  for (let i of range(a.length)) {
+    const aa = [...a];
+    const rr = [...r, ...aa.splice(i, 1)];
+    yield* permutations(aa, rr);
+  }
+}
+
+function getMatchedFamilyForVariationSequence(
+    familyList, baseCharacter, variationSelector) {
+  const variationSequence = baseCharacter + variationSelector;
+  // First try to find a match for the whole variation sequence.
+  if (variationSequenceFamilies.has(variationSequence)) {
+    const matchedFamily = variationSequenceFamilies.get(variationSequence);
+    if (familyList.includes(matchedFamily)) {
+      return matchedFamily;
+    }
+  }
+  // If failed, try to match only the base character from the
```

Comment:
Is this behavior actually specified in the spec?

Reply:
That's how I interpret [this statement](https://www.w3.org/TR/css-fonts-4/#cluster-matching:~:text=If%20no%20font%20on%20the%20system%20supports%20the%20full%20sequence%2C%20match%20the%20single%20character%20b%20using%20the%20normal%20procedure%20for%20matching%20single%20characters%20and%20ignore%20the%20variation%20selector.): 

> If no font on the system [supports](https://www.w3.org/TR/css-fonts-4/#support) the full sequence, match the single character b using the normal procedure for matching single characters and ignore the variation selector.